File: Backend/schema/resume_schema.py
from datetime import date
from pydantic import BaseModel, Field, EmailStr, field_validator
from typing import Annotated, Optional, Literal, List
import re

class Name(BaseModel):
    first_name: Annotated[str, Field(..., min_length=3, max_length=50, title="First Name", description="The First Name of the user", example="John")]
    middle_name: Annotated[Optional[str], Field(None, min_length=3, max_length=50, title="Middle Name", description="The Middle Name of the user", example="Michael")]
    last_name: Annotated[str, Field(..., min_length=3, max_length=100, title="Last Name", description="The Last Name of the user", example="Doe")]

    @field_validator('first_name')
    @classmethod
    def validate_name(cls, v: str) -> str:
        if not re.match(r"^[a-zA-Z\s]+$", v):
            raise ValueError("First Name can only contain letters and spaces")
        return v.strip()
    
    @field_validator('middle_name')
    @classmethod
    def validate_name(cls, v: str) -> str:
        if not re.match(r"^[a-zA-Z\s]+$", v):
            raise ValueError("Middle Name can only contain letters and spaces")
        return v.strip()

    @field_validator('last_name')
    @classmethod
    def validate_name(cls, v: str) -> str:
        if not re.match(r"^[a-zA-Z\s]+$", v):
            raise ValueError("Last Name can only contain letters and spaces")
        return v.strip()

class Location(BaseModel):
    address: Annotated[Optional[str], Field(None, min_length=5, max_length=200, description="Street address including house/building number and street name.", example="123 Main Street")]
    city: Annotated[str, Field(..., min_length=2, max_length=100, description="City or locality name.", example="New York")]
    state: Annotated[str, Field(..., min_length=2, max_length=100, description="State, province, or region.", example="NY")]
    country: Annotated[str, Field(..., min_length=2, max_length=100, description="Country name.", example="USA")]
    postal_code: Annotated[str, Field(..., min_length=3, max_length=12, description="Postal code or ZIP code.", example="10001")]

    @field_validator("city", "state", "country")
    @classmethod
    def validate_location_names(cls, v: str) -> str:
        if not re.match(r"^[A-Za-z\s]+$", v):
            raise ValueError("Only letters and spaces are allowed.")
        return v.strip()

    @field_validator("address")
    @classmethod
    def validate_address(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("Address cannot be empty.")
        return v.strip()

    @field_validator("postal_code")
    @classmethod
    def validate_postal_code(cls, v: str) -> str:
        # Supports common international postal code formats
        if not re.match(r"^[A-Za-z0-9\s\-]{3,12}$", v):
            raise ValueError("Invalid postal code format.")
        return v.strip().upper()

class PersonalInfo(BaseModel):
    name: Annotated[Name, Field(..., title="Full Name", description="The Full Name of the user", example={"first_name": "John", "middle_name": "Michael", "last_name": "Doe"})] 
    email: Annotated[EmailStr, Field(..., title="Email", description="The Contact Email address of the user", example="john.doe@example.com")]
    phone: Annotated[str, Field(..., title="Phone Number", description="Phone Number of the user", example="+1234567890")]
    location: Annotated[Location, Field(..., title="Location", description="Location of the User", example={"address": "456 Oak Avenue", "city": "Los Angeles", "state": "CA", "country": "USA", "postal_code": "90210"})]
    linkedin: Annotated[Optional[str], Field(None, title="LinkedIN", description="Linkedin URL of User", example="https://www.linkedin.com/in/username")]
    github: Annotated[Optional[str], Field(None, title="Github", description="Github URL of User",example="https://github.com/username")]
    portfolio: Annotated[Optional[str], Field(None, title="Portfolio", description="Portfolio URL of User", example="https://www.username.com")]

    @field_validator("linkedin")
    @classmethod
    def validate_url(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return v
        if not (v.startswith("https://www.linkedin.com/in/")):
            raise ValueError("URL must start with https://www.linkedin.com/in/")
        return v
    
    @field_validator("github")
    @classmethod
    def validate_url(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return v
        if not (v.startswith("https://github.com/")):
            raise ValueError("URL must start with https://github.com/")
        return v
    
    @field_validator("portfolio")
    @classmethod
    def validate_url(cls, v: Optional[str]) -> Optional[str]:
        if v is None:
            return v
        if not (v.startswith("http://") or v.startswith("https://")):
            raise ValueError("URL must start with http:// or https://")
        return v

    @field_validator('phone')
    @classmethod
    def validate_name(cls, v: str) -> str:
        if not re.match(r"^\+\d{1,3}\d{7,14}$" ,v):
            raise ValueError("Phone number must start with country code (e.g., +1) and contain only digits")
        return v.strip()

class Projects(BaseModel):
    project_name: Annotated[str, Field(..., max_length=150, title= "Project Name", description="Name of the project", example="E-commerce Website")]
    team_size: Annotated[int, Field(..., gt=0, title="Team Size", description="Number of team members involved in the project", example=4)]
    start_date: Annotated[date, Field(..., title="Start Date", description="Project start date in YYYY-MM-DD format", example="2021-01-01")]
    end_date: Annotated[Optional[date], Field(None, title="End Date", description="Project end date in YYYY-MM-DD format, or null if ongoing", example="2022-12-31")]
    project_url: Annotated[str, Field(..., title="Project URL", description="URL to the project repository or live site", example="https://github.com/username/e-commerce-website")]
    technologies_used: Annotated[List[str], Field(..., max_length=500, title="Technologies Used", description="Comma-separated list of technologies used in the project", example=["Python", "FastAPI"])]
    description: Annotated[str, Field(..., max_length=1000, title="Project Description", description="A brief description of the project", example="Developed a full-stack e-commerce website with user authentication and payment integration.")]
    
    @field_validator("end_date")
    @classmethod
    def validate_end_date(cls, v, info):
        start_date = info.data.get("start_date")

        # If both exist, end_date must be after start_date
        if v and start_date and v < start_date:
            raise ValueError("end_date cannot be before start_date")

        return v
    
class Experience(BaseModel):
    company: Annotated[str, Field(..., max_length=150, title="Company Name", description="Name of the company", example="Google")]
    designation: Annotated[str, Field(..., max_length=100, title="Designation", description="Job title or role held at the company", example="Software Engineer")]
    ctc: Annotated[float, Field(..., gt=0, title="CTC", description="Cost to Company (CTC) in INR", example=1200000.00)]
    location: Annotated[str, Field(..., title="Location", description="Location of the company", example="Bangalore")]
    start_date: Annotated[date, Field(..., title="Start Date", description="Employment start date in YYYY-MM-DD format", example="2020-01-01")]
    end_date: Annotated[Optional[date], Field(None, title="End Date", description="Employment end date in YYYY-MM-DD format, or null if ongoing", example="2022-12-31")]
    skills: Annotated[List[str], Field(default=[], title="Skills", description="List of skills acquired in this experience", example=["Python", "FastAPI"])]

    @field_validator("end_date")
    @classmethod
    def validate_end_date(cls, v, info):
        start_date = info.data.get("start_date")

        # If both exist, end_date must be after start_date
        if v and start_date and v < start_date:
            raise ValueError("end_date cannot be before start_date")

        return v
    
class Education(BaseModel):
    institution_name: Annotated[str, Field(..., max_length=150, title="Institution Name", description="Name of the educational institution", example="Stanford University")]
    degree: Annotated[str, Field(..., max_length=100, title="Degree", description="Degree or qualification obtained", example="Bachelor of Science")]
    field_of_study: Annotated[str, Field(..., max_length=100, title="Field of Study", description="Field of study or major", example="Computer Science")]
    start_date: Annotated[date, Field(..., title="Start Date", description="Education start date in YYYY-MM-DD format", example="2016-09-01")]
    end_date: Annotated[Optional[date], Field(None, title="End Date", description="Education end date in YYYY-MM-DD format, or null if ongoing", example="2020-06-30")]
    grade: Annotated[Optional[str], Field(None, max_length=20, title="Grade", description="Grade or percentage obtained", example="3.8/4.0")]

    @field_validator("end_date")
    @classmethod
    def validate_end_date(cls, v, info):
        start_date = info.data.get("start_date")

        # If both exist, end_date must be after start_date
        if v and start_date and v < start_date:
            raise ValueError("end_date cannot be before start_date")

        return v
    
class Certification(BaseModel):
    title: Annotated[str, Field(..., max_length=150, title="Certification Name", description="Name of the certification", example="AWS Certified Solutions Architect")]
    issuing_organization: Annotated[Optional[str], Field(None, max_length=150, title="Issuer", description="Name of the certification issuer", example="Amazon Web Services")]
    issue_date: Annotated[Optional[date], Field(None, title="Issue Date", description="Date when the certification was issued in YYYY-MM-DD format", example="2022-01-01")]
    skills: Annotated[List[str], Field(default=[], title="Skills", description="List of skills acquired from this certification", example=["Cloud Architecture", "DevOps"])]
    
class ResumeDetails(BaseModel):
    personal_info: Annotated[
        PersonalInfo, 
        Field(..., title="Personal Information", description="Personal information of the user", example={"first_name": "John", "last_name": "Doe", "email": "john.doe@example.com"})]
    skills: Annotated[
        List[str], 
        Field(default=[], title="Skills", description="List of skills possessed by the user", example=["Python", "Machine Learning"])]
    experience: Annotated[
        Optional[List[Experience]], 
        Field(default=[], title="Experience", description="List of work experiences", example=[{"company": "Google", "designation": "Software Engineer", "ctc": 1200000.00, "location_type": "bangalore", "start_date": "2020-01-01", "end_date": "2022-12-31", "skills": ["Python","FastAPI"]}])]
    education: Annotated[
        List[Education], 
        Field(default=[], title="Education", description="List of educational qualifications", example=[{"institution_name": "Stanford University", "degree": "Bachelor of Science", "field_of_study": "Computer Science", "start_date": "2016-09-01", "end_date": "2020-06-30", "grade": "3.8/4.0"}])]
    projects: Annotated[
        List[Projects], 
        Field(default=[], title="Projects", description="List of projects undertaken", example=[{"project_name": "E-commerce Website", "team_size": 5, "start_date": "2020-01-01", "end_date": "2022-12-31", "project_url": "https://github.com/johndoe/ecommerce", "technologies_used": ["React", "Node.js"], "description": "Developed a full-stack e-commerce website using React and Node.js"}])]
    certifications: Annotated[
        List[Certification], 
        Field(default=[], title="Certifications", description="List of certifications obtained", example=[{"title": "AWS Certified Solutions Architect", "issuing_organization": "Amazon Web Services", "issue_date": "2022-01-01", "skills": ["Cloud Architecture", "DevOps"]}])]
    technical_participation: Annotated[
        Optional[List[str]], 
        Field(None, max_length=500, title="Technical Participation", description="Description of technical participation in events, workshops, or competitions", example="Participated in Google Code Jam and Facebook Hacker Cup")]
    co_curricular: Annotated[
        Optional[List[str]], 
        Field(None, max_length=500, title="Co-Curricular Activities", description="Description of co-curricular activities", example="Member of the university coding club, participated in hackathons and coding competitions")]
    extra_curricular: Annotated[
        Optional[List[str]], 
        Field(None, max_length=500, title="Extra Curricular Activities", description="Description of extra curricular activities", example="Captain of the university basketball team")]
    achievements: Annotated[
        Optional[List[str]], 
        Field(None, max_length=500, title="Achievements", description="Description of achievements", example="Won first place in XYZ coding competition")]

File: Frontend/src/lib/resume-schema.ts
import { z } from "zod";

/* Zod schemas map to backend base_resume shape. */

export const personalInfoSchema = z.object({
  first_name: z
    .string()
    .min(3, "First Name must be at least 3 characters")
    .max(50, "First Name must be at most 50 characters")
    .regex(/^[a-zA-Z\s]+$/, "First Name can only contain letters and spaces"),
  middle_name: z
    .string()
    .optional()
    .default("")
    .refine(
      (val) => !val || (val.length >= 1 && val.length <= 50 && /^[a-zA-Z\s]+$/.test(val)),
      "Middle Name must be between 1 and 50 characters and only contain letters and spaces"
    ),
  last_name: z
    .string()
    .min(3, "Last Name must be at least 3 characters")
    .max(100, "Last Name must be at most 100 characters")
    .regex(/^[a-zA-Z\s]+$/, "Last Name can only contain letters and spaces"),
  email: z.string().email("Invalid email"),
  phone: z
    .string()
    .regex(/^\+\d{1,3}\d{7,14}$/, "Phone number must start with country code (e.g., +1) and contain only digits"),
  address: z
    .string()
    .optional()
    .default("")
    .refine(
      (val) => !val || (val.trim().length >= 5 && val.trim().length <= 200),
      "Address must be between 5 and 200 characters"
    ),
  city: z.string().optional().default(""),
  state: z.string().optional().default(""),
  country: z.string().optional().default(""),
  postal_code: z.string().optional().default(""),
  linkedin: z
    .string()
    .optional()
    .default("")
    .refine(
      (val) => !val || val.startsWith("https://www.linkedin.com/in/"),
      "URL must start with https://www.linkedin.com/in/"
    ),
  github: z
    .string()
    .optional()
    .default("")
    .refine(
      (val) => !val || val.startsWith("https://github.com/"),
      "URL must start with https://github.com/"
    ),
  portfolio: z
    .string()
    .optional()
    .default("")
    .refine(
      (val) => !val || val.startsWith("http://") || val.startsWith("https://"),
      "URL must start with http:// or https://"
    ),
}).superRefine((data, ctx) => {
  const hasAnyLocation = !!(
    data.address?.trim() ||
    data.city?.trim() ||
    data.state?.trim() ||
    data.country?.trim() ||
    data.postal_code?.trim()
  );

  if (hasAnyLocation) {
    if (!data.city || data.city.trim().length < 2 || data.city.trim().length > 100 || !/^[A-Za-z\s]+$/.test(data.city)) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "City is required, must be 2-100 characters, and contain only letters/spaces",
        path: ["city"],
      });
    }
    if (!data.state || data.state.trim().length < 2 || data.state.trim().length > 100 || !/^[A-Za-z\s]+$/.test(data.state)) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "State is required, must be 2-100 characters, and contain only letters/spaces",
        path: ["state"],
      });
    }
    if (!data.country || data.country.trim().length < 2 || data.country.trim().length > 100 || !/^[A-Za-z\s]+$/.test(data.country)) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "Country is required, must be 2-100 characters, and contain only letters/spaces",
        path: ["country"],
      });
    }
    if (!data.postal_code || !/^[A-Za-z0-9\s\-]{3,12}$/.test(data.postal_code.trim())) {
      ctx.addIssue({
        code: z.ZodIssueCode.custom,
        message: "Postal code must be 3-12 characters and alphanumeric",
        path: ["postal_code"],
      });
    }
  }
});

const notInFuture = (val: string | null | undefined): boolean => {
  if (!val) return true;
  const trimmed = val.trim();
  if (!trimmed) return true;
  const parts = trimmed.split("-");
  const year = parseInt(parts[0], 10);
  const month = parseInt(parts[1] || "1", 10) - 1;
  const dateObj = new Date(year, month, 1);
  const now = new Date();
  const currentYearMonth = new Date(now.getFullYear(), now.getMonth(), 1);
  return dateObj <= currentYearMonth;
};

export const experienceSchema = z
  .object({
    company: z.string().min(1, "Required").max(150, "Company name must be at most 150 characters"),
    designation: z.string().min(1, "Required").max(100, "Designation must be at most 100 characters"),
    ctc: z.string().refine((val) => {
      const num = parseFloat(val);
      return !isNaN(num) && num > 0;
    }, "CTC must be a positive number"),
    location: z.string().min(1, "Required"),
    start_date: z.string().min(1, "Required").refine(notInFuture, "Date cannot be in the future"),
    end_date: z.string().optional().default("").refine(notInFuture, "Date cannot be in the future"),
    description: z.string().min(1, "Required").max(1000, "Description must be at most 1000 characters"),
    skills: z.array(z.string()).default([]),
  })
  .refine(
    (data) => {
      if (data.start_date && data.end_date) {
        return new Date(data.end_date) >= new Date(data.start_date);
      }
      return true;
    },
    {
      message: "End date cannot be before start date",
      path: ["end_date"],
    }
  );

export const educationSchema = z
  .object({
    institution: z.string().min(1, "Required").max(150, "Institution name must be at most 150 characters"),
    degree: z.string().min(1, "Required").max(100, "Degree must be at most 100 characters"),
    field_of_study: z.string().min(1, "Required").max(100, "Field of study must be at most 100 characters"),
    start_date: z.string().min(1, "Required").refine(notInFuture, "Date cannot be in the future"),
    end_date: z.string().optional().default("").refine(notInFuture, "Date cannot be in the future"),
    grade: z.string().max(20, "Grade must be at most 20 characters").optional().default(""),
  })
  .refine(
    (data) => {
      if (data.start_date && data.end_date) {
        return new Date(data.end_date) >= new Date(data.start_date);
      }
      return true;
    },
    {
      message: "End date cannot be before start date",
      path: ["end_date"],
    }
  );

export const projectSchema = z
  .object({
    project_name: z.string().min(1, "Required").max(150, "Project name must be at most 150 characters"),
    team_size: z.string().refine((val) => {
      const num = parseInt(val, 10);
      return !isNaN(num) && num > 0;
    }, "Team size must be a positive integer"),
    project_url: z.string().url("Must be a valid URL"),
    technologies: z.array(z.string()).min(1, "At least one technology required").max(500, "Technologies must be at most 500 items"),
    description: z.string().min(1, "Required").max(1000, "Description must be at most 1000 characters"),
    start_date: z.string().min(1, "Required").refine(notInFuture, "Date cannot be in the future"),
    end_date: z.string().optional().default("").refine(notInFuture, "Date cannot be in the future"),
  })
  .refine(
    (data) => {
      if (data.start_date && data.end_date) {
        return new Date(data.end_date) >= new Date(data.start_date);
      }
      return true;
    },
    {
      message: "End date cannot be before start date",
      path: ["end_date"],
    }
  );

export const certificationSchema = z.object({
  title: z.string().min(1, "Required").max(150, "Title must be at most 150 characters"),
  issuer: z.string().max(150, "Issuer must be at most 150 characters").optional().default(""),
  issue_date: z.string().optional().default("").refine(notInFuture, "Date cannot be in the future"),
  skills: z.array(z.string()).default([]),
});

export const listItemSchema = z.object({
  title: z.string().min(1, "Required").max(200, "Title must be at most 200 characters"),
  description: z.string().max(300, "Description must be at most 300 characters").optional().default(""),
});

export const resumeSchema = z.object({
  personal_info: personalInfoSchema,
  skills: z.array(z.string()).default([]),
  experience: z.array(experienceSchema).default([]),
  education: z.array(educationSchema).default([]),
  projects: z.array(projectSchema).default([]),
  certifications: z.array(certificationSchema).default([]),
  achievements: z.array(listItemSchema).default([]),
  technical_participation: z.array(listItemSchema).default([]),
  co_curricular: z.array(listItemSchema).default([]),
  extra_curricular: z.array(listItemSchema).default([]),
});

export type Resume = z.infer<typeof resumeSchema>;
export type PersonalInfo = z.infer<typeof personalInfoSchema>;
export type Experience = z.infer<typeof experienceSchema>;
export type Education = z.infer<typeof educationSchema>;
export type Project = z.infer<typeof projectSchema>;
export type Certification = z.infer<typeof certificationSchema>;
export type ListItem = z.infer<typeof listItemSchema>;

export const emptyResume = (): Resume => ({
  personal_info: {
    first_name: "",
    middle_name: "",
    last_name: "",
    email: "",
    phone: "",
    address: "",
    city: "",
    state: "",
    country: "",
    postal_code: "",
    linkedin: "",
    github: "",
    portfolio: "",
  },
  skills: [],
  experience: [],
  education: [],
  projects: [],
  certifications: [],
  achievements: [],
  technical_participation: [],
  co_curricular: [],
  extra_curricular: [],
});

export function computeCompletion(r: Resume) {
  const checks = [
    !!(
      r.personal_info.first_name &&
      r.personal_info.email &&
      r.personal_info.phone
    ),
    r.education.length > 0,
    r.skills.length > 0,
    r.experience.length > 0,
    r.projects.length > 0,
    r.certifications.length > 0,
    r.achievements.length > 0,
  ];
  const done = checks.filter(Boolean).length;
  return {
    done,
    total: checks.length,
    pct: Math.round((done / checks.length) * 100),
    checks,
  };
}
